File: Code/1229. Meeting Scheduler.java
class Solution {
    public List<Integer> minAvailableDuration(int[][] slots1, int[][] slots2, int duration) {
        if (slots1 == null || slots2 == null){
            return new ArrayList<Integer>();
        }
        Arrays.sort(slots1,  (a, b) -> Integer.compare(a[0], b[0]));
        Arrays.sort(slots2,  (a, b) -> Integer.compare(a[0], b[0]));
        
        int i = 0;
        int j = 0;
        Integer[] resArray = new Integer[2];
        while (i < slots1.length && j < slots2.length){
            int[] left = slots1[i];
            int[] right = slots2[j];
            if (left[0] >= right[1]){
                j++;
                continue;
            }else if (right[0] >= left[1]){
                i++;
                continue;
            }
            

            if (Math.min(left[1], right[1]) - Math.max(left[0], right[0]) >= duration){
                resArray[0] = Math.max(left[0], right[0]);
                resArray[1] = resArray[0] + duration;
                return new ArrayList<Integer>(Arrays.asList(resArray));
            }else{
                if(left[1] >= right[1]){
                    j++;
                }else{
                    i++;
                }
            }
        }
        return new ArrayList<Integer>();
    }
}

File: Code/1229. Meeting Scheduler.py
class Solution:
    def minAvailableDuration(self, slots1: List[List[int]], slots2: List[List[int]], duration: int) -> List[int]:
        slots1.sort()
        slots2.sort()
        i = 0 
        j = 0
        while i < len(slots1) and j < len(slots2):
            first = slots1[i]
            second = slots2[j]
            if first[0] <= second[0]:
                # case 1
                if second[0] < first[1]:
                     # has interval
                    interval = min(first[1], second[1]) - second[0]
                    if interval >= duration:
                        return[second[0], second[0] + duration]
                    else:
                        if first[1] <= second[1]:
                            i += 1
                        else:
                            j += 1
                else:
                     # no interval
                    i += 1
            else:
                # case 2
                if first[0] < second[1]:
                    # has interval
                    interval = min(second[1], first[1]) - first[0]
                    if interval >= duration:
                        return [first[0], first[0] + duration]
                    else:
                        if second[1] <= first[1]:
                            j += 1
                        else:
                            i += 1
                else:
                    j += 1
        return []
                